src/main.cpp

```diff
@@ -6658,26 +6646,30 @@ CMutableTransaction CreateNewContextualCMutableTransaction(const Consensus::Para
     bool isOverwintered = consensusParams.NetworkUpgradeActive(nHeight, Consensus::UPGRADE_OVERWINTER);
     if (isOverwintered) {
         mtx.fOverwintered = true;
+        if (consensusParams.NetworkUpgradeActive(nHeight, Consensus::UPGRADE_SAPLING)) {
+            mtx.nVersionGroupId = SAPLING_VERSION_GROUP_ID;
+            mtx.nVersion = SAPLING_TX_VERSION;
+        } else {
+            mtx.nVersionGroupId = OVERWINTER_VERSION_GROUP_ID;
+            mtx.nVersion = OVERWINTER_TX_VERSION;
+        }
+
         mtx.nExpiryHeight = nHeight + (expiryDeltaArg ? expiryDeltaArg.get() : consensusParams.DefaultExpiryDelta(nHeight));
 
-        // Review: Should the following if() be checking the expiry delta rather than the transaction height?
+        // Review: is mtx.nExpiryHeight == 0 allowed? Is this coinbase only?
```

Comment:
Section 7.1:
> [Overwinter onward] If a transaction is not a coinbase transaction and its `nExpiryHeight` field is nonzero,
then it MUST NOT be mined at a block height greater than its `nExpiryHeight`.

In other words, `nExpiryHeight` is specified to be ignored for coinbase transactions, and in practice we do set it to zero for coinbase transactions.